test/coverlet.core.tests/Coverage/CoverageTest.Yield.cs

```diff
@@ -100,5 +100,65 @@ public void Yield_SingleWithSwitch()
                 File.Delete(path);
             }
         }
+
+        [Fact]
+        public void Yield_Three()
+        {
+            string path = Path.GetTempFileName();
+            try
+            {
+                FunctionExecutor.Run(async (string[] pathSerialize) =>
+                {
+                    CoveragePrepareResult coveragePrepareResult = await TestInstrumentationHelper.Run<Yield>(instance =>
+                    {
+                        foreach (var _ in instance.Three()) ;
+
+                        return Task.CompletedTask;
+                    }, persistPrepareResultToFile: pathSerialize[0]);
+                    return 0;
+                }, new string[] { path });
+
+                CoverageResult result = TestInstrumentationHelper.GetCoverageResult(path);
+
+                result.Document("Instrumentation.Yield.cs")
+                      .Method("System.Boolean Coverlet.Core.Samples.Tests.Yield/<Three>d__3::MoveNext()")
+                      .AssertLinesCovered((42, 1), (43, 1), (44, 1))
+                      .ExpectedTotalNumberOfBranches(0);
+            }
+            finally
+            {
+                File.Delete(path);
+            }
+        }
+
+        [Fact]
+        public void Yield_Enumerable()
+        {
+            string path = Path.GetTempFileName();
+            try
+            {
+                FunctionExecutor.Run(async (string[] pathSerialize) =>
+                {
+                    CoveragePrepareResult coveragePrepareResult = await TestInstrumentationHelper.Run<Yield>(instance =>
+                    {
+                        foreach (var _ in instance.Enumerable(new[] { "one", "two", "three", "four" })) ;
+
+                        return Task.CompletedTask;
+                    }, persistPrepareResultToFile: pathSerialize[0]);
+                    return 0;
+                }, new string[] { path });
+
+                CoverageResult result = TestInstrumentationHelper.GetCoverageResult(path);
+
+                result.Document("Instrumentation.Yield.cs")
+                      .Method("System.Boolean Coverlet.Core.Samples.Tests.Yield/<Enumerable>d__4::MoveNext()")
+                      .AssertLinesCovered((51, 4))
+                      .ExpectedTotalNumberOfBranches(1);
```

Comment:
Where is this branch?

Reply:
The body of `Yield.Enumerable()` is a `foreach` loop. I assumed that they involve a branch. But I'm actually not sure.

Reply:
Mmm maybe c# lies also here we should take a look at IL to understand if there is some branch, maybe an hidden null check

Reply:
Yup you are right. Here's the generated `MoveNext()`:

```
<Enumerable>d__0.MoveNext:
IL_0000:  ldarg.0     
IL_0001:  ldfld       UserQuery+Yield+<Enumerable>d__0.<>1__state
IL_0006:  stloc.1     
IL_0007:  ldloc.1     
IL_0008:  brfalse.s   IL_0012
IL_000A:  br.s        IL_000C
IL_000C:  ldloc.1     
IL_000D:  ldc.i4.1    
IL_000E:  beq.s       IL_0014
IL_0010:  br.s        IL_0016
IL_0012:  br.s        IL_001D
IL_0014:  br.s        IL_006A
IL_0016:  ldc.i4.0    
IL_0017:  stloc.0     
IL_0018:  leave       IL_00A1
IL_001D:  ldarg.0     
IL_001E:  ldc.i4.m1   
IL_001F:  stfld       UserQuery+Yield+<Enumerable>d__0.<>1__state
IL_0024:  nop         
IL_0025:  nop         
IL_0026:  ldarg.0     
IL_0027:  ldarg.0     
IL_0028:  ldfld       UserQuery+Yield+<Enumerable>d__0.ls
IL_002D:  callvirt    System.Collections.Generic.IEnumerable<System.String>.GetEnumerator
IL_0032:  stfld       UserQuery+Yield+<Enumerable>d__0.<>s__1
IL_0037:  ldarg.0     
IL_0038:  ldc.i4.s    FD 
IL_003A:  stfld       UserQuery+Yield+<Enumerable>d__0.<>1__state
IL_003F:  br.s        IL_007A
IL_0041:  ldarg.0     
IL_0042:  ldarg.0     
IL_0043:  ldfld       UserQuery+Yield+<Enumerable>d__0.<>s__1
IL_0048:  callvirt    System.Collections.Generic.IEnumerator<System.String>.get_Current
IL_004D:  stfld       UserQuery+Yield+<Enumerable>d__0.<item>5__2
IL_0052:  nop         
IL_0053:  ldarg.0     
IL_0054:  ldarg.0     
IL_0055:  ldfld       UserQuery+Yield+<Enumerable>d__0.<item>5__2
IL_005A:  stfld       UserQuery+Yield+<Enumerable>d__0.<>2__current
IL_005F:  ldarg.0     
IL_0060:  ldc.i4.1    
IL_0061:  stfld       UserQuery+Yield+<Enumerable>d__0.<>1__state
IL_0066:  ldc.i4.1    
IL_0067:  stloc.0     
IL_0068:  leave.s     IL_00A1
IL_006A:  ldarg.0     
IL_006B:  ldc.i4.s    FD 
IL_006D:  stfld       UserQuery+Yield+<Enumerable>d__0.<>1__state
IL_0072:  nop         
IL_0073:  ldarg.0     
IL_0074:  ldnull      
IL_0075:  stfld       UserQuery+Yield+<Enumerable>d__0.<item>5__2
IL_007A:  ldarg.0     
IL_007B:  ldfld       UserQuery+Yield+<Enumerable>d__0.<>s__1
IL_0080:  callvirt    System.Collections.IEnumerator.MoveNext
IL_0085:  brtrue.s    IL_0041
IL_0087:  ldarg.0     
IL_0088:  call        UserQuery+Yield+<Enumerable>d__0.<>m__Finally1
IL_008D:  nop         
IL_008E:  ldarg.0     
IL_008F:  ldnull      
IL_0090:  stfld       UserQuery+Yield+<Enumerable>d__0.<>s__1
IL_0095:  ldc.i4.0    
IL_0096:  stloc.0     
IL_0097:  leave.s     IL_00A1
IL_0099:  ldarg.0     
IL_009A:  call        UserQuery+Yield+<Enumerable>d__0.System.IDisposable.Dispose
IL_009F:  nop         
IL_00A0:  endfinally  
IL_00A1:  ldloc.0     
IL_00A2:  ret
```

`IL_0085` is a `brtrue.s` back up to `IL_0041` which seems to implement a loop like:

```csharp
do {
    // ...
} while (MoveNext())
```

Reply:
After having a closer look at this case it turned out to be a true nightmare. I don't see a way how to skip that branch from the `foreach` without the possibility of accidentally skipping a branch that actually was part of the user's code.

It might be that I'll be able to find some kind of pattern in the generated IL when I compare lots of different cases, but I'm not sure when I'll find the time to do that.

Reply:
For now it's ok...I had the same thought when I see that branch...too risky an btw that branch will be always covered if there is at least one element.